docs/public/components/testing.md

```diff
@@ -0,0 +1,203 @@
+To recap from the previous page, a component was created that compresses the payload before sending it to the Datadog backend. The component had two separate implementations.
+
+This is the component's interface:
+
+=== ":octicons-file-code-16: comp/compression/def/component.go"
+    ```go
+    type Component interface {
+        // Compress compresses the input data.
+        Compress([]byte) ([]byte, error)
+
+        // Decompress decompresses the input data.
+        Decompress([]byte) ([]byte, error)
+    }
+    ```
+
+Ensure the `Compress` and `Decompress` functions behave correctly.
+
+Writing tests for a component implementation follows the same rules as any other test in a Go project. See the [testing package](https://pkg.go.dev/testing) documentation for more information.
+
+For this example, write a test file for the `zstd` implementation. Create a new file named `component_test.go` in the `impl-zstd folder`. Inside the test file, initialize the component's dependencies, create a new component instance, and test the behavior.
+
+### Initialize the component's dependencies
+
+All components expect a `Requires` struct with all the necessary dependencies. To ensure a component instance can be created, create a `requires` instance.
+
+The `Requires` struct declares a dependency on the config component and the log component. The following code snippet shows how to create the `Require` struct:
+
+=== ":octicons-file-code-16: comp/compression/impl-zstd/component_test.go"
+    ```go
+    package implzstd
+    
+    import (
+      "testing"
+
+      configmock "github.com/DataDog/datadog-agent/comp/core/config/mock"
+      logmock "github.com/DataDog/datadog-agent/comp/core/log/mock"
+    )
+    
+    func TestCompress(t *testing.T) {
+      logComponent := configmock.New(t)
+      configComponent := logmock.New(t)
+      
+      requires := Requires{
+        Conf: configComponent,
+        Log: logComponent,
+      }
+      // [...]
+    }
+    ```
+    
+To create the log and config component, use their respective mocks. The [mock package](creating-components.md#the-mock-folder) was mentioned previously in the [Creating a Component page](creating-components.md).
+    
+
+### Testing the component's interface
+
+Now that the `Require` struct is created, an instance of the component can be created and its functionality tested:
+
+=== ":octicons-file-code-16: comp/compression/impl-zstd/component_test.go"
+    ```go
+    package implzstd
+    
+    import (
+      "testing"
+
+      configmock "github.com/DataDog/datadog-agent/comp/core/config/mock"
+      logmock "github.com/DataDog/datadog-agent/comp/core/log/mock"
+    )
+    
+    func TestCompress(t *testing.T) {
+      logComponent := configmock.New(t)
+      configComponent := logmock.New(t)
+      
+      requires := Requires{
+        Conf: configComponent,
+        Log: logComponent,
+      }
+      
+      provides := NewComponent(requires)
+      component := provides.Comp
+      
```

Comment:
The main idea is to not use Fx related functions inside the implementation folder. The new exposed public function `NewComponent` and the `Requires` struct, we no longer need to use FX to handle dependency injection.

Reply:
Discussed a bit via chat. The core idea is that it should be possible to turn a `impl` folder into a go module, and fx should be avoided in the `impl` folder so that users outside the Agent codebase can import it. Tests should be testing the implementation itself, so they should be using a mock. What we really need are easy ways to make mocks, which the `multi-impl` RFC did not go into detail on.

Fx specific tests (like those using `fxutil.TestRun` or `fxutil.TestOneShotCommand`) are only ever defined at the entry points (typically in the `cmd/` folder) and not in the component implementation folders.